### PR TITLE
[Issue-884] Implement hooks configuration for the GraphQL plugin

### PR DESCRIPTION
**Pull request by vecerek**
### What does this PR do?
This PR implements the execute hook for the graphql plugin based on #884. Its signature is the following:

```ts
type ExecuteHook = (span?: Span, context?: ExecutionContext, res?: ExecutionResult) => void
```

Links to definitions:

- [ExecutionContext](https://github.com/graphql/graphql-js/blob/c9850c1dab8811086e6fd8383508384e538b31fd/src/execution/execute.js#L89-L105)
- [ExecutionResult](https://github.com/graphql/graphql-js/blob/c9850c1dab8811086e6fd8383508384e538b31fd/src/execution/execute.js#L113)

cc: @rochdev

### Motivation

- #884 

### Plugin Checklist

- [x] Unit tests.
- [x] TypeScript [definitions][1].
- [x] TypeScript [tests][2].
- [ ] API [documentation][3].
- [x] CircleCI [jobs/workflows][4]. (Not Applicable)
- [x] Plugin is [exported][5]. (Not Applicable)

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Additional Notes

I wasn't able to update the API documentation for the plugin. It also serves a 404 page: https://github.com/DataDog/dd-trace-js/blob/master/docs/interfaces/plugins.graphql.html.

### CODE REVIEW

File: docs/test.ts [hunk 1]
```diff
@@ -116,7 +116,13 @@ const graphqlOptions = {
   depth: 2,
   variables: ({ foo, baz }) => ({ foo }),
   collapse: false,
-  signature: false
+  signature: false,
+  hooks: {
+    parse: (span, args) => {},
+    validate: (span, args) => {},
+    execute: (span, args, res) => {},
```
Comment (rochdev) [inline docs/test.ts:123]: Given the fundamental differences between these different hooks, I'd say to focus this PR only on `execute` for now and discuss the others separately.

File: index.d.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { ClientRequest, IncomingMessage, ServerResponse } from "http";
 import * as opentracing from "opentracing";
 import { SpanOptions } from "opentracing/lib/tracer";
+import { ExecutionArgs, ExecutionResult } from "graphql"
```
Comment (rochdev) [inline index.d.ts:4]: This doesn't work since the tracer doesn't depend on `graphql`. While unfortunate, I think the only solution is to copy the definitions.

File: index.d.ts [hunk 3]
```diff
@@ -796,6 +797,20 @@ declare namespace plugins {
...
+     * An object of optional callbacks to be executed during the respective
+     * phase of a GraphQL operation. Undefined callbacks default to a noop
+     * function.
+     * 
+     * @defaul {}
+     */
+    hooks?: {
+      parse?: (span: Span, args: { source: string }) => void;
+      validate?: (span: Span, args: { source: string }) => void;
+      execute?: (span: Span, args: ExecutionArgs, res: ExecutionResult) => void;
```
Comment (rochdev) [inline index.d.ts:811]: Should it be the `ExecutionContext` instead? It's closer to what is currently passed to other hooks.

File: packages/datadog-plugin-graphql/src/index.js [hunk 4]
```diff
@@ -34,8 +34,13 @@ function createWrapExecute (tracer, config, defaultFieldResolver) {
       contextValue._datadog_graphql = { source, span, fields: {} }
 
       return call(execute, span, this, [args], (err, res) => {
-        finishResolvers(contextValue)
-        finish(err || (res && res.errors && res.errors[0]), span)
+        finishResolvers(contextValue, config)
+
+        finish(err || (res && res.errors && res.errors[0]), span, undefined, (span) => config.hooks.execute(
```
Comment (rochdev) [inline packages/datadog-plugin-graphql/src/index.js:39]: Can this be rewritten in a way that doesn't require passing `undefined` arguments? I think it could be called right before calling `finish`.

File: index.d.ts (hunk 2 above)
Comment (vecerek) [reply index.d.ts:4]: I thought it was enough to have `graphql` in the dev dependencies to use the type definitions: https://github.com/DataDog/dd-trace-js/blob/master/package.json#L99.

File: index.d.ts (hunk 3 above)
Comment (vecerek) [reply index.d.ts:811]: I've chosen `ExecutionArgs` because it has a better match with what is returned by the `normalizeArgs` function.

```ts
// https://github.com/graphql/graphql-js/blob/157edcb4f7dfe9cb2acd3d82831430d983bbfbcf/src/execution/execute.d.ts#L29
export interface ExecutionContext {
  schema: GraphQLSchema;
  fragments: { [key: string]: FragmentDefinitionNode }; // not returned by normalizeArgs
  rootValue: any;
  contextValue: any;
  operation: OperationDefinitionNode; // not returned by normalizeArgs (it's operationName instead)
  variableValues: { [key: string]: any };
  fieldResolver: GraphQLFieldResolver<any, any>;
  errors: Array<GraphQLError>; // not returned by normalizeArgs
}

// https://github.com/graphql/graphql-js/blob/157edcb4f7dfe9cb2acd3d82831430d983bbfbcf/src/execution/execute.d.ts#L51
export type ExecutionArgs = {
  schema: GraphQLSchema;
  document: DocumentNode;
  rootValue?: any;
  contextValue?: any;
  variableValues?: Maybe<{ [key: string]: any }>;
  operationName?: Maybe<string>;
  fieldResolver?: Maybe<GraphQLFieldResolver<any, any>>;
  typeResolver?: Maybe<GraphQLTypeResolver<any, any>>; // not returned by normalizeArgs but it's an optional property at least
};

// https://github.com/DataDog/dd-trace-js/blob/master/packages/datadog-plugin-graphql/src/index.js#L211-L219
function normalizeArgs (args) {
  if (args.length === 1) {
    return args[0]
  }

  return {
    schema: args[0],
    document: args[1],
    rootValue: args[2],
    contextValue: args[3],
    variableValues: args[4],
    operationName: args[5],
    fieldResolver: args[6]
  }
};
```

File: packages/datadog-plugin-graphql/src/index.js (hunk 4 above)
Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js:39]: I'm passing the hook down to `finish` this way for one reason. I think the hook should be called after the plugin adds any possible error tags to the span but before the `finish` is called on the span.
```js
function finish (error, span, finishTime, hook) {
  if (error) {
    span.addTags({
      'error.type': error.name,
      'error.msg': error.message,
      'error.stack': error.stack
    })
  }

  // Call hook before span#finish is called but after errors have possibly been added
  if (hook) hook(span)

  span.finish(finishTime)
}
```

This would possibly enable the client to remove the error tag if necessary. There's no API currently for removing tags but I think it might be possible to go around it with:
```js
span.addTags({
  'error.type': undefined,
  'error.msg': undefined,
  'error.stack': undefined
})
```

The use case for this would be, as discussed in the issue, to determine whether a graphql operation failed due to client or server errors. Client errors should not result in the execution span to be marked as an error.

Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js:39]: We could get rid of these `undefined` values if we changed the order of params:
```js
function finish (error, span, finishTime, hook) { ... } // old
function finish (error, span, hook, finishTime) { ... } // new
```
However, undefined values might appear in other places then (for which there's no hook defined, yet).

File: index.d.ts (hunk 2 above)
Comment (vecerek) [reply index.d.ts:4]: Ahh, I see now. The version of `graphql` in the dev dependencies is relatively old. That version doesn't export any TS definitions. Would it make sense to update the dependency?

Comment (vecerek) [reply index.d.ts:4]: Copying over the types would be a relatively large effort as well. Skimming over the definition files of `graphql` I can see it would require to copy over the following files/fragments of files:
- https://github.com/graphql/graphql-js/blob/157edcb4f7dfe9cb2acd3d82831430d983bbfbcf/src/type/schema.d.ts
- https://github.com/graphql/graphql-js/blob/157edcb4f7dfe9cb2acd3d82831430d983bbfbcf/src/language/ast.d.ts#L201-L297
- parts of https://github.com/graphql/graphql-js/blob/157edcb4f7dfe9cb2acd3d82831430d983bbfbcf/src/type/definition.d.ts
- https://github.com/graphql/graphql-js/blob/157edcb4f7dfe9cb2acd3d82831430d983bbfbcf/src/error/GraphQLError.d.ts
- https://github.com/graphql/graphql-js/blob/157edcb4f7dfe9cb2acd3d82831430d983bbfbcf/src/tsutils/Maybe.d.ts

The more I'm looking at it, the less convinced I am the copy approach is worth it (mostly due to maintenance implications) 😞

File: docs/test.ts (hunk 1 above)
Comment (vecerek) [reply docs/test.ts:121]: Makes sense. I reduced the scope in https://github.com/DataDog/dd-trace-js/pull/909/commits/e3edb8fb355820ca2bf2ee419cdbc75b59038315.

File: index.d.ts (hunk 2 above)
Comment (rochdev) [reply index.d.ts:4]: The main issue is that the definitions are exposed outside the library, so devDependencies would be missing and the definitions would not work even if it's up-to-date in development.

If it's too much work to copy the definitions, I'd say to simply use `any` for now. This is a bit of an unfortunate limitation of TypeScript which pretty much requires all libraries to be monorepos exposing types as a separate package to work properly.

File: index.d.ts (hunk 3 above)
Comment (rochdev) [reply index.d.ts:809]: Correct, but I'm still not sure it's the right approach.

If I compare this to HTTP for example, it would look like this:

```js
const req = http.get(options, (res) => {})
```

* `ExecutionArgs` is equivalent to `options`. It's provided only once by the user when the operation is executed.
* `ExecutionContext` is equivalent to `req`. It represents the actual request and is available throughout the library.
* `ExecutionResult` is equivalent to `res`. It represents the response and is available at the end of the operation.

In that sense, I'd lean toward `ExecutionContext` unless there is a strong argument for `ExecutionArgs`.

File: packages/datadog-plugin-graphql/src/index.js (hunk 4 above)
Comment (rochdev) [reply packages/datadog-plugin-graphql/src/index.js:39]: That makes sense, I didn't think about the error. Maybe then it would be better to split `finish` and add the error separately? Then it could look like this:

```js
addError(err)
executeHook(span, context, res)
span.finish()
```

Although I'm not sure the best way to prevent the error is to set every field to `undefined` as it feels a bit hacky. I'll take a look to see if it could be done otherwise with the current API.

Comment (rochdev) [reply packages/datadog-plugin-graphql/src/index.js:39]: Ok I think the best syntax for now would probably be to rewrite how the error is added to use `span.setTag('error', error)` instead. This is a new syntax that was not supported when the plugin was written. Then, to remove the error all that will be needed is to do `span.setTag('error', false)` or `span.setTag('error', 0)`.

File: index.d.ts (hunk 3 above)
Comment (vecerek) [reply index.d.ts:809]: When I'm looking at the definition of the two types, I don't see a huge conceptual difference. They both have 8 properties, out of which 3 differ (`ExecutionArgs` -> `ExecutionContext`):

1. `document` -> `fragments`,
2. `operationName` -> `operation`,
3. `typeResolver` -> `errors`.

In the first case, `ExecutionArgs` provides more information because fragments can be derived from the document using the following code:
```js
const fragments = document && document.definitions.reduce((acc, d) => {
  if (d.kind !== 'FragmentDefinition') return acc
  return {
    ...acc,
    [d.name.value]: d
  }
}, {})
```

In the second case, `ExecutionContext` provides more information because `operationName` can be derived from `operation`.

In the last case, `ExecutionArgs` provides more information because the typeresolver is missing from `ExecutionContext` and `errors` can be derived from `ExecutionResult`.

It may be desirable to provide as much information as possible to the hook. That could be achieved with `ExecutionArgs` where `operationName` would be replaced with `operation`, unless I'm missing something 🙂

File: packages/datadog-plugin-graphql/src/index.js (hunk 4 above)
Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js:39]: > Maybe then it would be better to split finish and add the error separately? 

Should it be split for all wrappers that use `finish` or just in the case of the execution wrapper?

File: index.d.ts (hunk 2 above)
Comment (vecerek) [reply index.d.ts]: I've changed it to use `any` instead: https://github.com/DataDog/dd-trace-js/pull/909/commits/a1554b505e55c7d5f0f02b94f87790352a8e5e54.

File: packages/datadog-plugin-graphql/src/index.js (hunk 4 above)
Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js:39]: Replaced `addTags` with `setTag('error', error)`: https://github.com/DataDog/dd-trace-js/pull/909/commits/f6c57e27066b00732f50d2c8cd82819c61ecc689.

Comment (adamthalhammer) [reply packages/datadog-plugin-graphql/src/index.js:39]: As suggested in https://github.com/DataDog/dd-trace-js/issues/884#issuecomment-612568623, I think it's worth considering what the default behaviour should be and how to supply it.

Proposal: move the `setTag('error', error)` into the default (currently noop) `execute` callback. Consequences:
* consumers can opt-out of the current behaviour by supplying their own callback
* consumers don't need to un-do the `setTag` call at runtime
* the default `execute` callback can serve as documentation for how to use the hook

File: index.d.ts (hunk 3 above)
Comment (adamthalhammer) [reply index.d.ts]: Appreciate all your hard work and detailed analysis on this topic, @vecerek and @rochdev!

Disclaimer: I am not intimately familiar with the two types in question. But I hope my perspective as an outsider will be of some value. While I appreciate the aim of providing as much information as possible to the hook, doing so comes at the expense of some conceptual clarity. As @rochdev points out, each of the types serves a particular purpose and has a particular meaning with well-established patterns around where and how they are intended to be used.

Therefore, I would strongly encourage the use of `ExecutionContext`.

Comment (rochdev) [reply index.d.ts]: @vecerek I still think `ExecutionContext` makes the most sense for relevance and availability, even if maybe not for the overall raw data. It's also the safest option so that if we move the instrumentation deeper into the `grapghql` module we still have access to the `ExecutionContext` since it's what's available within the framework.

Again, all of the above also applies to HTTP since some information from `options` maybe not be available in `req`. I think this is an acceptable limitation since usually the information is still available but in a different form, for example a URL string vs a URL object, or a document vs fragments.

Of course it's possible that I'm completely missing some use cases, but at that point I think we could always add the `ExecutionArgs` as an additional parameter. The same could be said for HTTP if `options` is ever needed in a hook.

File: packages/datadog-plugin-graphql/src/index.js (hunk 4 above)
Comment (rochdev) [reply packages/datadog-plugin-graphql/src/index.js:39]: > Should it be split for all wrappers that use finish or just in the case of the execution wrapper?

I would say for now whatever results in the least amount of code. When other hooks are implemented, part of this will probably be refactored anyway so for now let's keep it simple.

> consumers can opt-out of the current behaviour by supplying their own callback

In general the hooks are general-purpose functions that are meant to augment the spans from the plugins, but not the auto-instrumentation. In that sense I think a config option would be more adequate for this specific use case.

File: index.d.ts (hunk 3 above)
Comment (vecerek) [reply index.d.ts]: Let's do that then 🙂Addressed in this commit:  https://github.com/DataDog/dd-trace-js/pull/909/commits/29fadaf2f23df64418cb50c53dfd43889593f729

File: packages/datadog-plugin-graphql/src/index.js (hunk 4 above)
Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js]: Addressed in https://github.com/DataDog/dd-trace-js/pull/909/commits/9af35067071f369cb00e964903fb1f29926d40b3.

**Comment by vecerek** [comment]
> Proposal: move the setTag('error', error) into the default (currently noop) execute callback.

@adamthalhammer I think this would make a lot of sense. Unfortunately, this means we would have to pass the error down to the hook. When the execution function is wrapped, it is provided a callback which is called with an `err` and a `res` object. `err` has precedence over `res.errors[0]` when setting the error:

```ts
return call(execute, span, this, [args], (err, res) => {
  finishResolvers(contextValue, config)

  setError(span, err || (res && res.errors && res.errors[0]))
  config.hooks.execute(span, getExecutionContext(args, res), res)
  finish(span)
})
```

@rochdev what do you think about setting the error as the default execute hook?

**Comment by rochdev** [comment]
> what do you think about setting the error as the default execute hook?

I don't think adding a hook should alter the default behavior, especially for use cases that are completely unrelated to errors in which case that would be an unexpected loss of functionality. I do agree however that it's a bit weird to unset and already set error, although the overhead is minimal since the error already exists. I think this should be done as a separate effort for another configuration option outside of this PR. It could be for example a list of error types, or a function to validate what is considered an error.

**Comment by vecerek** [comment]
@rochdev I believe I addressed all outstanding comments. May I ask you to re-review the PR? 🙂

File: docs/test.ts [hunk 5]
```diff
@@ -116,7 +116,10 @@ const graphqlOptions = {
   depth: 2,
   variables: ({ foo, baz }) => ({ foo }),
   collapse: false,
-  signature: false
+  signature: false,
+  hooks: {
+    execute: (span, args, res) => {},
```
Comment (rochdev) [inline docs/test.ts:121]: `args` should be renamed to `context` or equivalent to avoid confusion.

File: index.d.ts [hunk 6]
```diff
@@ -796,6 +796,17 @@ declare namespace plugins {
...
+
+    /**
+     * An object of optional callbacks to be executed during the respective
+     * phase of a GraphQL operation. Undefined callbacks default to a noop
+     * function.
+     * 
+     * @default {}
+     */
+    hooks?: {
+      execute?: (span: Span, args: any, res: any) => void;
```
Comment (rochdev) [inline index.d.ts:808]: All of the arguments be optional.

File: packages/datadog-plugin-graphql/src/index.js [hunk 7]
```diff
@@ -16,7 +16,7 @@ function createWrapExecute (tracer, config, defaultFieldResolver) {
       const source = document && document._datadog_source
       const fieldResolver = args.fieldResolver || defaultFieldResolver
       const contextValue = args.contextValue = args.contextValue || {}
-      const operation = getOperation(document, args.operationName)
+      const operation = args.operation = getOperation(document, args.operationName)
```
Comment (rochdev) [inline packages/datadog-plugin-graphql/src/index.js:19]: Why is this needed?

File: packages/datadog-plugin-graphql/src/index.js [hunk 8]
```diff
@@ -318,25 +323,24 @@ function startResolveSpan (tracer, config, childOf, path, info, contextValue) {
...
+    span.setTag('error', error)
   }
+}
 
+function finish (span, finishTime) {
   span.finish(finishTime)
 }
 
-function finishResolvers (contextValue) {
+function finishResolvers (contextValue, config) {
```
Comment (rochdev) [inline packages/datadog-plugin-graphql/src/index.js:336]: It doesn't look like `config` is used.

File: packages/datadog-plugin-graphql/src/index.js [hunk 9]
```diff
@@ -392,11 +396,39 @@ function getOperation (document, operationName) {
...
+  return document.definitions.reduce((acc, d) => {
+    if (d.kind !== 'FragmentDefinition') return acc
+    return {
+      ...acc,
+      [d.name.value]: d
+    }
+  }, {})
+}
+
+function getExecutionContext (args, res) {
```
Comment (rochdev) [inline packages/datadog-plugin-graphql/src/index.js:413]: This should not be necessary since `graphql` already does this. Using what is provided by `graphql` will also avoid duplicating this logic.

File: packages/datadog-plugin-graphql/test/index.spec.js [hunk 10]
```diff
@@ -1290,6 +1290,96 @@ describe('Plugin', () => {
         })
       })
 
+      describe('with hooks configuration', () => {
+        const config = {
+          hooks: {
+            parse: sinon.spy((span, args) => {}),
```
Comment (rochdev) [inline packages/datadog-plugin-graphql/test/index.spec.js:1296]: These no longer exist except `execute`.

File: packages/datadog-plugin-graphql/src/index.js (hunk 9 above)
Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js:413]: The plugin does not currently depend on graphql. So, I tried to move the dependency from dev to regular. When I did that and imported `buildExecutionContext` from `graphql/execution/execute`, it broke the tests:

```
Error: Cannot use GraphQLScalarType "String" from another module or realm.

Ensure that there is only one instance of "graphql" in the node_modules
directory. If different versions of "graphql" are the dependencies of other
relied on modules, use "resolutions" to ensure only one version is installed.

https://yarnpkg.com/en/docs/selective-version-resolutions

Duplicate "graphql" modules cannot be used at the same time since different
versions may have different capabilities and behavior. The data from one
version used in the function from another could produce confusing and
spurious results.
```

The tests are run against 2 different versions of graphql. I'm not familiar with these kinds of setups and I suspect that could be the root cause of the failure. Can you help me resolve the issue?

File: packages/datadog-plugin-graphql/src/index.js (hunk 7 above)
Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js]: It isn't. I assigned it to args, so I can pass it down to `getExecutionContext`. That way it wouldn't have to be recomputed. I removed this, since we'll want to use `graphql`'s native `buildExecutionContext` instead.

File: packages/datadog-plugin-graphql/src/index.js (hunk 9 above)
Comment (rochdev) [reply packages/datadog-plugin-graphql/src/index.js:413]: Since it's exported from the module, it should be possible to grab it from the plugin directly. It should be available at `execute.buildExecutionContext`. In that sense it's not required to require it externally since it's already passed to the plugin.

However, doing so means you would end up with a new execution context that is different than the one provided to the rest of the library internally. It would also mean that the execution context is computed twice, which has an overhead.

Unfortunately, it doesn't look like the internal execution context is available anywhere outside of the `execution/execute.js` file, meaning there is no way to intercept it. Since the execution args may be missing some fields, I think at this point the only possible option will be to pass a custom object that includes any missing field that were added by the tracer. In my opinion the important ones would be everything from the args except the field resolver and type resolver.

In short, because it is unfortunately impossible to use the execution context, we will have to use the args, but augment it when fields are missing to ensure there are references to these fields in the plugin. Back to args it is 😄

Comment (rochdev) [reply packages/datadog-plugin-graphql/src/index.js:413]: Actually it looks like the missing pieces are still available eventually on the `info` object passed to resolvers. So in theory is would be possible to combine `args` and `info` to reconstruct the context. I wonder if this should be done at the execute hook level, or if it should be up to the user to combine 2 hooks (execute + resolve) to get the information.

Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js:413]: > ...we will have to use the args, but augment it when fields are missing...

Do I understand correctly that what we want now is an ExecutionArgs-like object with most of its fields being required?

```ts
// Original ExecutionArgs type
export type ExecutionArgs = {
  schema: GraphQLSchema,
  document: DocumentNode,
  rootValue?: mixed,
  contextValue?: mixed,
  variableValues?: ?{ +[variable: string]: mixed, ... },
  operationName?: ?string,
  fieldResolver?: ?GraphQLFieldResolver<any, any>,
  typeResolver?: ?GraphQLTypeResolver<any, any>,
};

// What we want:
export type ExecutionArgsButBetter = {
  schema: GraphQLSchema,
  document: DocumentNode,
  rootValue: mixed,
  contextValue: mixed,
  variableValues: { [variable: string]: mixed, ... },
  operationName: string,
  fieldResolver?: ?GraphQLFieldResolver<any, any>,
  typeResolver?: ?GraphQLTypeResolver<any, any>,
};
```
> So in theory is would be possible to combine args and info to reconstruct the context. I wonder if this should be done at the execute hook level

How could the context be reconstructed in the execute hook? The `info` object is only available to the wrapped field resolver. It gets passed into the resolver somewhere outside of the scope of the execution wrapper.

> or if it should be up to the user to combine 2 hooks (execute + resolve) to get the information.

I'm not sure I understand how that could be achieved. Could you give an example?

Comment (rochdev) [reply packages/datadog-plugin-graphql/src/index.js:413]: > Do I understand correctly that what we want now is an ExecutionArgs-like object with most of its fields being required?

Correct. However, for now I would probably only include common fields that are mutable within the operation to be safe. The type definition would then look something like this:

```ts
export type ExecutionContext = {
  rootValue: any,
  contextValue: any,
  variableValues: any
};
```

Most of the other fields are more difficult to interact with or have limited relevance. I prefer to keep it limited for now since it's still unclear if the args is really the only option.

> How could the context be reconstructed in the execute hook? The info object is only available to the wrapped field resolver. It gets passed into the resolver somewhere outside of the scope of the execution wrapper.

That could be a problem, but not necessarily depending on how it's handled internally. The root resolver should have access to the execution context, but I don't know if there are any guarantees on the presence of at least one resolver. That's something to look into, but since it's more involved I think it's best to try to do it outside the scope of this PR if possible.

> I'm not sure I understand how that could be achieved. Could you give an example?

The following `ExecutionContext` fields are available on the `ExecutionArgs` provided by the user:

* schema
* rootValue (may be undefined)
* contextValue (may be undefined)
* variableValues (may be undefined)
* fieldResolver (may be undefined but is not needed)
* typeResolver (may be undefined but is not needed)

The following `ExecutionContext` fields are available on the `GraphQLResolveInfo` passed to resolvers:

* schema
* operation
* fragments
* rootValue
* variableValues
* operation

By combining the fields from both, it's possible to create an object that is identical to the internal `ExecutionContext` without any additional processing. However, if this is not handled by the hook implementations, it means a user would theoretically have to combine an `execute` hook and `resolve` hook to get all of this information in a single place.

Sorry if this whole thing is getting a bit out of hand. I expected the hooks to be easier to implement with the assumption that the context was exposed by `graphql`.

Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js]: Thank you, @rochdev. I still don't understand the part about combining the fields from 2 different hooks. Nevertheless, I tried to address the requested change in https://github.com/DataDog/dd-trace-js/pull/909/commits/eab4f7f352f7959c8778b356a08af981df1b3efc. Let me know if that matches what you described in one of the comments above 🙂

Comment (adamthalhammer) [reply packages/datadog-plugin-graphql/src/index.js]: Thank you @vecerek for persevering on this PR, even though it turned out to be harder than expected ❤️

Comment (rochdev) [reply packages/datadog-plugin-graphql/src/index.js]: @vecerek It definitely matches what I had in mind. However I now realize that `rootValue` can be undefined and `variableValues` is different between `ExecutionArgs` and `ExecutionContext`. At this point I have to capitulate and agree that `ExecutionArgs` is the only way forward. I still think `ExecutionContext` would have been the superior option but it's way too complicated to support a similar object. For the type definition, I would say to do similar to the current `ExecutionContext` type but renamed and with all the keys that are on the actual `ExecutionArgs`, with the same required/optional flag but with everything as type `any`.

Sorry for the runaround. This should be a small change and then it can be merged.

Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js]: Addressed in https://github.com/DataDog/dd-trace-js/pull/909/commits/f8dbcf9ed23f359ffaa6e15f47214702aa9777a2. I hope I didn't forget about anything, unit tests are passing, shouldn't be too bad 😄

Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js]: Should I squash the commits, or would you prefer to leave them for historical reasons? 🙂

Comment (rochdev) [reply packages/datadog-plugin-graphql/src/index.js]: I'll squash the commits when merging to master, so it's your preference if you want to keep the history in the PR.

Comment (vecerek) [reply packages/datadog-plugin-graphql/src/index.js]: Let's keep it then.

**Comment by rochdev** [comment]
@kobecow You should be able to use a hook to detect whether there is an error and override the tag to remove the error like so:

```js
tracer.use('graphql', {
  hooks: {
    execute (span, args, res) {
      // detect from args/res if the request is valid then call span.setTag('error', false)
    }
  }
})
```
